src/storage/database.rs

```diff
@@ -0,0 +1,1483 @@
+    }
+
+    #[test]
+    fn cache_insert_empty_success() {
+        set_test_db_path();
+        let conn = init_database().unwrap();
+
+        let stmt_result = conn.prepare("SELECT * FROM cache;");
+        assert!(stmt_result.is_ok());
+        let mut stmt = stmt_result.unwrap();
+        let query_result = stmt.query_map([], |row| {
+            let path_str = row.get_unwrap::<usize, String>(3).to_owned();
+
+            Ok(CacheRow {
+                id: row.get_unwrap(0),
+                encoded_text: row.get_unwrap(1),
+                decoded_text: row.get_unwrap(2),
+                path: serde_json::from_str(&path_str).unwrap_or_default(),
+                successful: row.get_unwrap(4),
+                execution_time_ms: row.get_unwrap(5),
+                timestamp: row.get_unwrap(6),
+            })
+        });
+        assert!(query_result.is_ok());
+        let empty_rows = query_result.unwrap();
+        assert_eq!(empty_rows.count(), 0);
+    }
+
+    #[test]
+    fn cache_insert_success() {
+        set_test_db_path();
+        let conn = init_database().unwrap();
+
+        let encoded_text = String::from("aGVsbG8gd29ybGQK");
+        let decoded_text = String::from("hello world");
+
+        let (_mock_crack_result, mut expected_cache_row, cache_entry) =
+            generate_cache_row(1, &encoded_text, &decoded_text);
+        let row_result = insert_cache(&cache_entry);
+        assert!(row_result.is_ok());
+        assert_eq!(row_result.unwrap(), 1);
+
+        let stmt_result = conn.prepare("SELECT * FROM cache;");
+        let mut stmt = stmt_result.unwrap();
+        let query_result = stmt.query_map([], |row| {
+            let path_str = row.get_unwrap::<usize, String>(3).to_owned();
+
+            Ok(CacheRow {
+                id: row.get_unwrap(0),
+                encoded_text: row.get_unwrap(1),
+                decoded_text: row.get_unwrap(2),
+                path: serde_json::from_str(&path_str).unwrap_or_default(),
+                successful: row.get_unwrap(4),
+                execution_time_ms: row.get_unwrap(5),
+                timestamp: row.get_unwrap(6),
+            })
+        });
+        assert!(query_result.is_ok());
+        let cache_row: CacheRow = query_result.unwrap().next().unwrap().unwrap();
+        expected_cache_row.timestamp = cache_row.timestamp.clone();
+        assert_eq!(cache_row, expected_cache_row);
+    }
+
+    #[test]
+    fn cache_insert_2_success() {
+        set_test_db_path();
+        let conn = init_database().unwrap();
+
+        let encoded_text_1 = String::from("aGVsbG8gd29ybGQK");
+        let decoded_text_1 = String::from("hello world");
+
+        let (_mock_crack_result_1, mut expected_cache_row_1, cache_entry_1) =
+            generate_cache_row(1, &encoded_text_1, &decoded_text_1);
+        let row_result = insert_cache(&cache_entry_1);
+        assert!(row_result.is_ok());
+        assert_eq!(row_result.unwrap(), 1);
```

Comment:
i personally prefer 1 assert per function test because it lets nextest multi-thread it, but i guess this is not multi-threaded so its okay here